File: figures/src/python/classes/uncertainty_abstention_curve.py
#!/usr/bin/env python3
"""Class-based Python renderer for uncertainty-abstention curves."""

from __future__ import annotations

import csv
from pathlib import Path
from typing import Any

from matplotlib.figure import Figure

from figures.src.python.common import (
    REPO_ROOT,
    add_panel_labels,
    apply_publication_layout,
    configure_matplotlib,
    export_csv,
    mm_to_inches,
    project_resources,
    resolve_font_stack,
    source_data_mapping,
    validate_common_contract,
    write_manifest,
)


TARGET_RISK = 0.08
TARGET_COVERAGE = 0.80


def _load_curve_rows(path: Path) -> list[dict[str, Any]]:
    with path.open("r", encoding="utf-8", newline="") as handle:
        reader = csv.DictReader(handle)
        return [
            {
                "model": row["model"],
                "display_order": int(row["display_order"]),
                "coverage": float(row["coverage"]),
                "risk": float(row["risk"]),
                "risk_lower": float(row["risk_lower"]),
                "risk_upper": float(row["risk_upper"]),
                "operating_point": row["operating_point"],
            }
            for row in reader
        ]


def _load_summary_rows(path: Path) -> list[dict[str, Any]]:
    with path.open("r", encoding="utf-8", newline="") as handle:
        reader = csv.DictReader(handle)
        rows = [
            {
                "model": row["model"],
                "display_order": int(row["display_order"]),
                "risk_at_full_coverage": float(row["risk_at_full_coverage"]),
                "risk_at_80_coverage": float(row["risk_at_80_coverage"]),
                "coverage_at_target_risk": float(row["coverage_at_target_risk"]),
                "abstained_fraction_at_target": float(row["abstained_fraction_at_target"]),
                "label_model": row["label_model"],
            }
            for row in reader
        ]
    return sorted(rows, key=lambda item: item["display_order"])


def build_source_data(
    spec: dict[str, Any],
    curve_rows: list[dict[str, Any]],
    summary_rows: list[dict[str, Any]],
) -> tuple[list[dict[str, Any]], list[dict[str, Any]]]:
    ordered_curves = sorted(curve_rows, key=lambda item: (item["display_order"], item["coverage"]))
    ordered_summary = sorted(summary_rows, key=lambda item: item["display_order"])
    mapping = source_data_mapping(spec)
    export_csv(
        REPO_ROOT / mapping["a"],
        ordered_curves,
        [
            "model",
            "display_order",
            "coverage",
            "risk",
            "risk_lower",
            "risk_upper",
            "operating_point",
        ],
    )
    export_csv(
        REPO_ROOT / mapping["b"],
        ordered_summary,
        [
            "model",
            "display_order",
            "risk_at_full_coverage",
            "risk_at_80_coverage",
            "coverage_at_target_risk",
            "abstained_fraction_at_target",
            "label_model",
        ],
    )
    return ordered_curves, ordered_summary


def _style_maps(theme: dict[str, Any], summary_rows: list[dict[str, Any]]) -> dict[str, dict[str, Any]]:
    colors = [
        theme["palette"]["categorical"][0],
        theme["palette"]["categorical"][1],
        theme["palette"]["categorical"][2],
    ]
    line_styles = ["solid", "--", "-."]
    markers = ["o", "s", "D"]
    style_map: dict[str, dict[str, Any]] = {}
    for index, row in enumerate(summary_rows):
        style_map[str(row["model"])] = {
            "color": colors[index % len(colors)],
            "linestyle": line_styles[index % len(line_styles)],
            "marker": markers[index % len(markers)],
        }
    return style_map


def _label_offsets() -> dict[str, tuple[float, float]]:
    return {
        "Foundation model": (0.018, -0.010),
        "Hybrid GNN": (0.018, 0.006),
        "CNN baseline": (0.018, 0.012),
    }


def create_figure(spec_path: Path) -> Figure:
    spec, theme, font_policy, profile, class_entry = project_resources(spec_path)
    validate_common_contract(spec, theme, font_policy, profile, class_entry)
    curve_rows = _load_curve_rows(REPO_ROOT / str(spec["data_inputs"][0]))
    summary_rows = _load_summary_rows(REPO_ROOT / str(spec["data_inputs"][1]))
    panel_a_rows, panel_b_rows = build_source_data(spec, curve_rows, summary_rows)
    resolved_font = resolve_font_stack(font_policy)
    configure_matplotlib(theme, font_policy, resolved_font)

    import matplotlib.pyplot as plt
    from matplotlib.ticker import PercentFormatter

    fig, axes = plt.subplots(
        1,
        2,
        figsize=(
            mm_to_inches(float(spec["size"]["width_mm"])),
            mm_to_inches(float(spec["size"]["height_mm"])),
        ),
        gridspec_kw={"width_ratios": [1.15, 0.95]},
        constrained_layout=bool(theme["layout"]["constrained"]),
    )
    apply_publication_layout(fig, theme)
    style_map = _style_maps(theme, panel_b_rows)
    offsets = _label_offsets()

    panel_a = axes[0]
    for summary in panel_b_rows:
        model = str(summary["model"])
        model_rows = sorted(
            [row for row in panel_a_rows if str(row["model"]) == model],
            key=lambda item: float(item["coverage"]),
        )
        x_values = [float(row["coverage"]) for row in model_rows]
        y_values = [float(row["risk"]) for row in model_rows]
        lower = [float(row["risk_lower"]) for row in model_rows]
        upper = [float(row["risk_upper"]) for row in model_rows]
        panel_a.fill_between(
            x_values,
            lower,
            upper,
            color=style_map[model]["color"],
            alpha=0.13,
            linewidth=0.0,
        )
        panel_a.plot(
            x_values,
            y_values,
            color=style_map[model]["color"],
            linestyle=style_map[model]["linestyle"],
            linewidth=1.8,
        )
        op_row = next(row for row in model_rows if str(row["operating_point"]) == "yes")
        panel_a.scatter(
            float(op_row["coverage"]),
            float(op_row["risk"]),
            s=30,
            color=style_map[model]["color"],
            marker=style_map[model]["marker"],
            edgecolor=theme["palette"]["neutral"][0],
            linewidth=0.5,
            zorder=4,
        )
        dx, dy = offsets[model]
        panel_a.text(
            float(op_row["coverage"]) + dx,
            float(op_row["risk"]) + dy,
            model,
            fontsize=theme["typography"]["annotation_font_size_pt"],
            color=theme["palette"]["neutral"][0],
            ha="left",
            va="center",
        )

    panel_a.axhline(
        TARGET_RISK,
        color=theme["palette"]["neutral"][1],
        linestyle=":",
        linewidth=1.0,
    )
    panel_a.axvline(
        TARGET_COVERAGE,
        color=theme["palette"]["neutral"][2],
        linestyle="--",
        linewidth=0.8,
    )
    panel_a.text(
        0.50,
        TARGET_RISK + 0.006,
        "8% target risk",
        ha="left",
        va="bottom",
        fontsize=theme["typography"]["annotation_font_size_pt"],
        color=theme["palette"]["neutral"][1],
    )
    panel_a.text(
        TARGET_COVERAGE + 0.012,
        0.185,
        "80% coverage\noperating points",
        ha="left",
        va="top",
        fontsize=theme["typography"]["annotation_font_size_pt"],
        color=theme["palette"]["neutral"][1],
    )
    panel_a.set_title("Coverage-risk curve", loc="left", pad=3)
    panel_a.set_xlabel("Coverage retained")
    panel_a.set_ylabel("Observed risk")
    panel_a.set_xlim(0.48, 1.02)
    panel_a.set_ylim(0.0, 0.21)
    panel_a.xaxis.set_major_formatter(PercentFormatter(xmax=1.0, decimals=0))
    panel_a.yaxis.set_major_formatter(PercentFormatter(xmax=1.0, decimals=0))
    panel_a.spines["top"].set_visible(False)
    panel_a.spines["right"].set_visible(False)
    panel_a.grid(color=theme["palette"]["neutral"][2], linewidth=0.45, alpha=0.45)

    panel_b = axes[1]
    ordered_summary = list(reversed(panel_b_rows))
    y_positions = list(range(len(ordered_summary)))
    coverage_values = [float(row["coverage_at_target_risk"]) for row in ordered_summary]
    panel_b.barh(
        y_positions,
        coverage_values,
        color=[style_map[str(row["model"])]["color"] for row in ordered_summary],
        edgecolor=theme["palette"]["neutral"][0],
        linewidth=0.45,
        alpha=0.9,
    )
    panel_b.axvline(
        TARGET_COVERAGE,
        color=theme["palette"]["neutral"][1],
        linestyle="--",
        linewidth=0.8,
    )
    panel_b.set_yticks(y_positions)
    panel_b.set_yticklabels([str(row["model"]) for row in ordered_summary])
    panel_b.set_xlabel("Coverage at risk <= 8%")
    panel_b.set_title("Target-risk coverage", loc="left", pad=3)
    panel_b.set_xlim(0.0, 1.0)
    panel_b.xaxis.set_major_formatter(PercentFormatter(xmax=1.0, decimals=0))
    panel_b.spines["top"].set_visible(False)
    panel_b.spines["right"].set_visible(False)
    panel_b.spines["left"].set_visible(False)
    panel_b.grid(axis="x", color=theme["palette"]["neutral"][2], linewidth=0.6)
    for y_pos, row in zip(y_positions, ordered_summary):
        panel_b.text(
            float(row["coverage_at_target_risk"]) + 0.015,
            y_pos,
            f"{row['coverage_at_target_risk'] * 100:.0f}% retained\nrisk@80% {row['risk_at_80_coverage'] * 100:.1f}%",
            ha="left",
            va="center",
            fontsize=theme["typography"]["annotation_font_size_pt"],
            color=theme["palette"]["neutral"][0],
        )
    panel_b.text(
        TARGET_COVERAGE,
        -0.58,
        "80% target coverage",
        ha="center",
        va="top",
        fontsize=theme["typography"]["annotation_font_size_pt"],
        color=theme["palette"]["neutral"][1],
    )

    add_panel_labels(spec, theme, [panel_a, panel_b])
    fig.text(
        0.01,
        0.965,
        spec["title"],
        ha="left",
        va="top",
        fontsize=theme["typography"]["title_font_size_pt"],
        fontweight="bold",
        color=theme["palette"]["neutral"][0],
    )
    return fig


def build_figure(spec_path: Path) -> dict[str, str]:
    spec, theme, font_policy, profile, class_entry = project_resources(spec_path)
    validate_common_contract(spec, theme, font_policy, profile, class_entry)
    resolved_font = resolve_font_stack(font_policy)
    configure_matplotlib(theme, font_policy, resolved_font)
    fig = create_figure(spec_path)

    output_dir = REPO_ROOT / str(spec["renderers"]["python"]["output_dir"])
    output_dir.mkdir(parents=True, exist_ok=True)
    stem = spec["figure_id"]
    outputs = {
        "svg": str((output_dir / f"{stem}.svg").relative_to(REPO_ROOT)),
        "pdf": str((output_dir / f"{stem}.pdf").relative_to(REPO_ROOT)),
        "png": str((output_dir / f"{stem}.png").relative_to(REPO_ROOT)),
    }
    fig.savefig(REPO_ROOT / outputs["svg"], format="svg", transparent=False)
    fig.savefig(REPO_ROOT / outputs["pdf"], format="pdf", transparent=False)
    fig.savefig(
        REPO_ROOT / outputs["png"],
        format="png",
        dpi=int(profile["preview_dpi"]),
        transparent=False,
    )

    curve_rows = _load_curve_rows(REPO_ROOT / str(spec["data_inputs"][0]))
    summary_rows = _load_summary_rows(REPO_ROOT / str(spec["data_inputs"][1]))
    write_manifest(
        spec,
        profile,
        resolved_font,
        "python",
        spec_path,
        outputs,
        [
            "coverage_risk_curve_panel",
            "uncertainty_band_for_risk",
            "operating_point_markers_and_direct_labels",
            "target_risk_reference",
            "selective_prediction_summary_panel",
            "vector_first_export",
        ],
        {
            "panel_count": 2,
            "reference_line_count": 3,
            "annotation_count": 2 + len(summary_rows) + sum(1 for row in curve_rows if row["operating_point"] == "yes"),
            "operating_point_count": sum(1 for row in curve_rows if row["operating_point"] == "yes"),
            "uncertainty_band_count": len(summary_rows),
        },
    )

    import matplotlib.pyplot as plt

    plt.close(fig)
    return outputs


if __name__ == "__main__":
    build_figure(Path("figures/specs/figure_13_uncertainty_abstention_curve.yml"))

File: figures/src/r/classes/uncertainty_abstention_curve.R
repo_root <- getOption("manuscript.figure.repo_root")
if (is.null(repo_root)) {
  stop("manuscript.figure.repo_root option is required")
}
source(file.path(repo_root, "figures", "src", "r", "common.R"))

target_risk <- 0.08
target_coverage <- 0.80

load_curve_rows <- function(path) {
  rows <- read.csv(path, stringsAsFactors = FALSE, check.names = FALSE)
  rows$model <- as.character(rows$model)
  rows$operating_point <- as.character(rows$operating_point)
  rows <- rows[order(rows$display_order, rows$coverage), ]
  rows
}

load_summary_rows <- function(path) {
  rows <- read.csv(path, stringsAsFactors = FALSE, check.names = FALSE)
  rows$model <- as.character(rows$model)
  rows$label_model <- as.character(rows$label_model)
  rows <- rows[order(rows$display_order), ]
  rows
}

build_source_data <- function(spec, curve_rows, summary_rows) {
  mapping <- source_data_mapping(spec)
  write_source_csv(
    file.path(repo_root, mapping$a),
    curve_rows,
    c(
      "model",
      "display_order",
      "coverage",
      "risk",
      "risk_lower",
      "risk_upper",
      "operating_point"
    )
  )
  write_source_csv(
    file.path(repo_root, mapping$b),
    summary_rows,
    c(
      "model",
      "display_order",
      "risk_at_full_coverage",
      "risk_at_80_coverage",
      "coverage_at_target_risk",
      "abstained_fraction_at_target",
      "label_model"
    )
  )
  list(curves = curve_rows, summary = summary_rows)
}

style_maps <- function(theme, summary_rows) {
  colors <- c(
    theme$palette$categorical[[1]],
    theme$palette$categorical[[2]],
    theme$palette$categorical[[3]]
  )
  line_types <- c("solid", "dashed", "dotdash")
  shapes <- c(16, 15, 18)
  data.frame(
    model = summary_rows$model,
    color = colors[seq_len(nrow(summary_rows))],
    line_type = line_types[seq_len(nrow(summary_rows))],
    shape = shapes[seq_len(nrow(summary_rows))],
    stringsAsFactors = FALSE
  )
}

label_offsets <- function() {
  data.frame(
    model = c("Foundation model", "Hybrid GNN", "CNN baseline"),
    dx = c(0.018, 0.018, 0.018),
    dy = c(-0.010, 0.006, 0.012),
    stringsAsFactors = FALSE
  )
}

create_plot <- function(spec_path) {
  spec <- load_yaml(spec_path)
  theme <- load_yaml(file.path(repo_root, "figures/config/project_theme.yml"))
  font_policy <- load_yaml(file.path(repo_root, "figures/config/font_policy.yml"))
  profiles <- load_yaml(file.path(repo_root, "figures/config/venue_profiles.yml"))
  registry <- load_yaml(file.path(repo_root, "figures/registry/classes.yml"))$classes
  profile <- profiles$profiles[[spec$target_profile]]
  class_entry <- registry[[spec$class_id]]
  validate_common_contract(spec, theme, font_policy, profile, class_entry)
  resolved_font <- resolve_font(font_policy)

  curve_rows <- load_curve_rows(file.path(repo_root, spec$data_inputs[[1]]))
  summary_rows <- load_summary_rows(file.path(repo_root, spec$data_inputs[[2]]))
  source_rows <- build_source_data(spec, curve_rows, summary_rows)
  styles <- style_maps(theme, source_rows$summary)
  offsets <- label_offsets()

  curve_plot_rows <- merge(source_rows$curves, styles, by = "model", sort = FALSE)
  curve_labels <- merge(
    curve_plot_rows[curve_plot_rows$operating_point == "yes", ],
    offsets,
    by = "model",
    sort = FALSE
  )
  curve_labels$label_x <- curve_labels$coverage + curve_labels$dx
  curve_labels$label_y <- curve_labels$risk + curve_labels$dy

  plot_theme <- build_theme(theme, resolved_font) +
    ggplot2::theme(
      legend.position = "none",
      plot.margin = ggplot2::margin(8, 18, 10, 8)
    )

  plot_a <- ggplot2::ggplot(
    curve_plot_rows,
    ggplot2::aes(x = coverage, y = risk, color = model, linetype = model, group = model)
  ) +
    ggplot2::geom_ribbon(
      ggplot2::aes(ymin = risk_lower, ymax = risk_upper, fill = model),
      alpha = 0.13,
      linewidth = 0,
      show.legend = FALSE
    ) +
    ggplot2::geom_line(linewidth = 0.8, show.legend = FALSE) +
    ggplot2::geom_hline(yintercept = target_risk, linetype = "dotted", linewidth = 0.4, color = theme$palette$neutral[[2]]) +
    ggplot2::geom_vline(xintercept = target_coverage, linetype = "dashed", linewidth = 0.35, color = theme$palette$neutral[[3]]) +
    ggplot2::geom_point(
      data = curve_plot_rows[curve_plot_rows$operating_point == "yes", ],
      ggplot2::aes(shape = model),
      size = 2.2,
      stroke = 0.35,
      fill = "white",
      show.legend = FALSE
    ) +
    ggplot2::geom_text(
      data = curve_labels,
      ggplot2::aes(x = label_x, y = label_y, label = model),
      inherit.aes = FALSE,
      hjust = 0,
      size = as.numeric(theme$typography$annotation_font_size_pt) / 2.845,
      color = theme$palette$neutral[[1]]
    ) +
    ggplot2::annotate(
      "text",
      x = 0.50,
      y = target_risk + 0.006,
      label = "8% target risk",
      hjust = 0,
      vjust = 0,
      size = as.numeric(theme$typography$annotation_font_size_pt) / 2.845,
      color = theme$palette$neutral[[2]]
    ) +
    ggplot2::annotate(
      "text",
      x = target_coverage + 0.012,
      y = 0.185,
      label = "80% coverage\noperating points",
      hjust = 0,
      vjust = 1,
      size = as.numeric(theme$typography$annotation_font_size_pt) / 2.845,
      color = theme$palette$neutral[[2]]
    ) +
    ggplot2::scale_color_manual(values = stats::setNames(styles$color, styles$model)) +
    ggplot2::scale_fill_manual(values = stats::setNames(styles$color, styles$model)) +
    ggplot2::scale_linetype_manual(values = stats::setNames(styles$line_type, styles$model)) +
    ggplot2::scale_shape_manual(values = stats::setNames(styles$shape, styles$model)) +
    ggplot2::scale_x_continuous(labels = scales::percent_format(accuracy = 1)) +
    ggplot2::scale_y_continuous(labels = scales::percent_format(accuracy = 1)) +
    ggplot2::labs(title = "Coverage-risk curve", x = "Coverage retained", y = "Observed risk") +
    ggplot2::coord_cartesian(xlim = c(0.48, 1.02), ylim = c(0, 0.21), clip = "off") +
    plot_theme

  support_rows <- source_rows$summary[nrow(source_rows$summary):1, ]
  support_rows$model <- factor(support_rows$model, levels = support_rows$model)
  support_rows$label_value <- sprintf(
    "%.0f%% retained\nrisk@80%% %.1f%%",
    support_rows$coverage_at_target_risk * 100,
    support_rows$risk_at_80_coverage * 100
  )

  plot_b <- ggplot2::ggplot(support_rows, ggplot2::aes(x = coverage_at_target_risk, y = model, fill = model)) +
    ggplot2::geom_vline(xintercept = target_coverage, linewidth = 0.4, linetype = "dashed", color = theme$palette$neutral[[2]]) +
    ggplot2::geom_col(width = 0.68, color = theme$palette$neutral[[1]], linewidth = 0.35, alpha = 0.9, show.legend = FALSE) +
    ggplot2::geom_text(
      ggplot2::aes(label = label_value, x = coverage_at_target_risk + 0.015),
      hjust = 0,
      size = as.numeric(theme$typography$annotation_font_size_pt) / 2.845,
      color = theme$palette$neutral[[1]]
    ) +
    ggplot2::annotate(
      "text",
      x = target_coverage,
      y = 0.42,
      label = "80% target coverage",
      hjust = 0.5,
      vjust = 1,
      size = as.numeric(theme$typography$annotation_font_size_pt) / 2.845,
      color = theme$palette$neutral[[2]]
    ) +
    ggplot2::scale_fill_manual(values = stats::setNames(styles$color, styles$model)) +
    ggplot2::scale_x_continuous(labels = scales::percent_format(accuracy = 1)) +
    ggplot2::labs(title = "Target-risk coverage", x = "Coverage at risk <= 8%", y = NULL) +
    ggplot2::coord_cartesian(xlim = c(0, 1), clip = "off") +
    plot_theme +
    ggplot2::theme(
      axis.line.y = ggplot2::element_blank(),
      axis.ticks.y = ggplot2::element_blank()
    )

  figure_plot <- (plot_a + plot_b + patchwork::plot_layout(ncol = 2, widths = c(1.15, 0.95))) +
    patchwork::plot_annotation(
      title = spec$title,
      tag_levels = "a"
    ) &
    ggplot2::theme(
      plot.title = ggplot2::element_text(
        size = as.numeric(theme$typography$title_font_size_pt),
        family = resolved_font$family,
        face = "bold",
        hjust = 0
      ),
      plot.tag = ggplot2::element_text(
        size = as.numeric(theme$panel_labels$font_size_pt),
        face = theme$panel_labels$font_weight,
        family = resolved_font$family
      )
    )

  list(
    plot = figure_plot,
    spec = spec,
    theme = theme,
    profile = profile,
    resolved_font = resolved_font,
    curves = source_rows$curves,
    summary = source_rows$summary
  )
}

build_figure <- function(spec_path) {
  payload <- create_plot(spec_path)
  spec <- payload$spec
  theme <- payload$theme
  profile <- payload$profile
  resolved_font <- payload$resolved_font
  figure_plot <- payload$plot
  output_dir <- file.path(repo_root, spec$renderers$r$output_dir)
  dir.create(output_dir, recursive = TRUE, showWarnings = FALSE)
  stem <- spec$figure_id
  svg_path <- file.path(output_dir, sprintf("%s.svg", stem))
  pdf_path <- file.path(output_dir, sprintf("%s.pdf", stem))
  png_path <- file.path(output_dir, sprintf("%s.png", stem))
  width_in <- mm_to_inches(as.numeric(spec$size$width_mm))
  height_in <- mm_to_inches(as.numeric(spec$size$height_mm))

  svglite::svglite(file = svg_path, width = width_in, height = height_in, bg = theme$export_defaults$background, system_fonts = list(sans = resolved_font$family))
  print(figure_plot)
  grDevices::dev.off()
  grDevices::cairo_pdf(file = pdf_path, width = width_in, height = height_in, bg = theme$export_defaults$background, family = resolved_font$family)
  print(figure_plot)
  grDevices::dev.off()
  ragg::agg_png(filename = png_path, width = width_in, height = height_in, units = "in", res = as.numeric(profile$preview_dpi), background = theme$export_defaults$background)
  print(figure_plot)
  grDevices::dev.off()

  outputs <- list(
    svg = sub(paste0("^", repo_root, "/"), "", svg_path),
    pdf = sub(paste0("^", repo_root, "/"), "", pdf_path),
    png = sub(paste0("^", repo_root, "/"), "", png_path)
  )
  write_manifest(
    repo_root,
    spec,
    profile,
    resolved_font,
    spec_path,
    "r",
    outputs,
    list(
      "coverage_risk_curve_panel",
      "uncertainty_band_for_risk",
      "operating_point_markers_and_direct_labels",
      "target_risk_reference",
      "selective_prediction_summary_panel",
      "vector_first_export"
    ),
    list(
      panel_count = 2,
      reference_line_count = 3,
      annotation_count = 2 + nrow(payload$summary) + sum(payload$curves$operating_point == "yes"),
      operating_point_count = sum(payload$curves$operating_point == "yes"),
      uncertainty_band_count = nrow(payload$summary)
    )
  )
  outputs
}
